### PR TITLE
Fix Codex PDF continuation serialization (Fixes #2608)

### LINKED ISSUE #2608: Codex PDF reads send unsupported file input and break the agent loop
## Summary

Reading a PDF with the read_file tool while using the Codex provider can terminate the agent loop with a 400 response:

    API Error: Client error: Missing required parameter: 'input[362].content[0]'. (Status: 400)

The tool reports success, but the continuation request contains a PDF content item that the Codex backend cannot accept. This is a loop-breaking provider compatibility bug rather than a PDF rendering or tool-execution failure.

The same path also exposes a provider-independent serialization defect: filenames are discarded by read_file before the OpenAI Responses request is built, even though inline file_data input requires a filename.

## Reproduction

1. Select the built-in codex provider alias and a Codex model, for example the gpt56solhigh profile using gpt-5.6-sol.
2. Ask the model to inspect one or more local PDF files.
3. The model invokes read_file and the tool UI reports entries such as:

       Read pdf file: pdf-studio/test-artifacts/semantic-corpus/ai-tools/candidate.pdf

4. When LLxprt submits the tool results back to the model, the Codex endpoint returns HTTP 400 and the loop stops.

This was observed with four parallel PDF reads, but the specific error is request-schema validation rather than an aggregate-size error. The active continuation path combines those files into one media-bearing user turn; it does not multiply every PDF once per tool response.

## Expected behavior

LLxprt must never send a media content type that the selected provider/backend or model does not accept.

- For an endpoint that supports native PDF file input, read_file output should be serialized as a valid file content item with the source filename preserved.
- For Codex, which currently does not expose PDF/file message input, LLxprt should replace the PDF media block with an explicit model-visible text notice instead of sending input_file. The notice must state that the PDF was not read and identify the file when possible, so the model can choose a supported conversion path rather than claim it inspected the document.
- Unsupported media must not abort the agent loop.

### 1. Codex currently does not advertise message-level PDF/file input

The built-in alias targets the private Codex Responses endpoint:

    https://chatgpt.com/backend-api/codex/responses

Current upstream Codex source defines model-visible ContentItem variants for input_text, input_image, input_audio, and output_text. It has no input_file variant. Current UserInput and Python SDK input unions similarly expose text, image, local image, audio in Rust, skills, and mentions, but no generic file/PDF turn input.

Upstream evidence:

- ContentItem source, with no InputFile variant:
  https://github.com/openai/codex/blob/main/codex-rs/protocol/src/models.rs#L700-L718
- UserInput source, with no file/PDF variant:
  https://github.com/openai/codex/blob/main/codex-rs/protocol/src/user_input.rs#L11-L55
- Codex Python SDK inputs expose text/images/skills/mentions, not generic files:
  https://github.com/openai/codex/blob/main/sdk/python/src/openai_codex/_inputs.py
- Upstream PDF support request remains open:
  https://github.com/openai/codex/issues/1797
- Upstream modality gating models support as text/image capability metadata and blocks unsupported attachments before submission:
  https://github.com/openai/codex/pull/10271
- Upstream structured tool-output work states that Codex supports text and image only for now, although the general Responses API supports files:
  https://github.com/openai/codex/pull/10567

The public OpenAI Responses API does support PDF input_file content, including GPT-5.6, but that does not imply parity with the private Codex endpoint:

https://developers.openai.com/api/docs/guides/file-inputs

### 2. LLxprt drops the PDF filename before serialization

The read path is:

1. packages/core/src/utils/fileUtils.ts processMediaFile reads bytes and creates legacy inlineData with data and mimeType only.
2. packages/core/src/utils/generateContentResponseUtilities.ts converts inlineData to MediaBlock with mimeType, data, and encoding only.
3. packages/providers/src/openai-responses/OpenAIResponsesInputBuilder.ts convertMediaBlock emits input_file with file_data and conditionally emits filename only when MediaBlock.filename already exists.

A local end-to-end payload reproduction using the repository PDF fixture produced:

    {
      "role": "user",
      "content": [
        {
          "type": "input_file",
          "file_data": "data:application/pdf;base64,..."
        }
      ]
    }

There is no filename. Public Responses inline Base64 file examples include filename, and omitting it is known to produce the same generic missing content-item error.

By contrast, the OpenAI Chat Completions request builder already supplies a fallback filename of document.pdf.

### 3. The active builder has no backend/model media capability input

ResponsesInputBuildContext currently carries reasoning and output-limiting concerns, but no resolved media capability. It always maps application/pdf to input_file, including when the target is the Codex backend.

The repository already has a model capability shape with a pdf boolean, and provider alias ephemerals/model defaults can carry request-time behavior overrides. Neither is currently consumed by the Responses input builder.

### 4. Tests cover idealized media blocks rather than the failing production path

- The file utility PDF test checks MIME type and Base64 data but not filename propagation.
- Detailed Responses PDF tests manually construct MediaBlock values with filenames and target the older buildResponsesInputFromContent helper.
- The active OpenAIResponsesInputBuilder tests cover tool pairing but do not cover read_file PDF output or Codex capability gating.

As a result, focused media and builder tests pass while the real continuation payload is invalid or unsupported.

## Proposed design

Use explicit request-time media capabilities at the provider serialization boundary rather than hard-coding model names or teaching read_file about providers.

### Capability resolution

1. Add a model behavior/ephemeral for native PDF input, suggested key:

       media.pdf.enabled

2. Resolve it using the established invocation/model behavior/settings precedence and pass the resolved value into ResponsesInputBuildContext.
3. Set media.pdf.enabled to false in the built-in codex provider alias. This expresses an endpoint capability and applies consistently to all Codex models until upstream support is demonstrated.
4. Keep the public OpenAI Responses behavior enabled by default or derive it from the existing hydrated model capabilities.pdf value where that capability is reliably available. Endpoint-specific alias behavior must be able to override model-registry claims because the Codex backend is not the public Responses API.

This follows the same architectural direction as upstream Codex input modality gating: decide support before request serialization and degrade safely rather than relying on server rejection.

### Provider serialization

At OpenAIResponsesInputBuilder.convertMediaBlock:

- When the PDF capability is enabled, emit a valid input_file item and always include a non-empty filename.
- Preserve the actual source basename end-to-end from read_file when possible.
- Use a deterministic fallback such as document.pdf only for legacy/history media that truly lacks a filename.
- When the PDF capability is disabled, emit input_text using the existing unsupported-media placeholder infrastructure, but make the wording unambiguous that the PDF content was not read. Include the filename/path if available and suggest extracting text or rendering pages to images with an available tool.
- Do not silently drop the media block and do not report successful PDF understanding.

### Size validation

The current local read limit is 20 MB per file, while the public Responses API documents a 50 MB combined per-request file limit. Add provider-side aggregate validation for native file inputs or a clear preflight failure/placeholder so four locally valid PDFs cannot produce an opaque remote size error.

## Test-first implementation plan

1. Add a behavioral regression test that starts with realistic read_file PDF output with no pre-populated MediaBlock filename and asserts the exact active OpenAIResponsesInputBuilder request shape.
2. Add a Codex-mode/provider-capability test proving PDF tool output becomes model-visible input_text, contains an explicit cannot-read-PDF message, contains no input_file part, and keeps all tool call/output pairing valid.
3. Add a public OpenAI Responses test proving enabled PDF support emits input_file with a data URI and non-empty filename.
4. Add filename propagation tests at processMediaFile and legacyPartsToBlocks boundaries so the real source basename survives into MediaBlock.
5. Add tests for legacy/history PDF blocks without filenames and verify the deterministic fallback.
6. Add multiple-parallel-read coverage proving PDFs are not duplicated across tool outputs and capability handling is applied once per media block.
7. Add aggregate-size boundary tests for native file-capable Responses requests.
8. Update codex.config tests to pin media.pdf.enabled=false and verify alias ephemerals reach request construction.
9. Run the complete project verification suite after implementation.

Tests should assert observable request and continuation behavior rather than mocking only private helper calls.

## Acceptance criteria

- [ ] The Codex provider never sends input_file for PDF read_file output while native PDF capability is disabled.
- [ ] The model receives a clear text notice that the PDF was not read, with its filename/path when available, and the agent loop continues.
- [ ] The Codex alias explicitly declares PDF input unsupported through a request-time capability/ephemeral rather than a model-name conditional.
- [ ] Public OpenAI Responses requests that support PDFs include valid file_data and a non-empty filename.
- [ ] The real PDF source filename is preserved from read_file through neutral history into provider serialization.
- [ ] Legacy media lacking a filename receives a deterministic safe fallback when native PDF input is enabled.
- [ ] Parallel PDF tool results are serialized once each without cross-tool duplication.
- [ ] Oversized aggregate file input is handled locally with a clear error or fallback rather than an opaque provider 400.
- [ ] Regression tests exercise the active OpenAIResponsesInputBuilder and a realistic tool-output continuation path.
- [ ] Existing image/media behavior and tool call/output pairing remain unchanged.

## Scope notes

This is related to the provider media work in #1615, but it is not a duplicate. That issue addressed media being silently dropped. This report covers a later path where PDF media is retained, then serialized into a malformed and Codex-unsupported content type that aborts the loop.

### PR DESCRIPTION
**Pull request by acoliver**
## TLDR

Fixes Codex PDF tool-result continuations so unsupported native PDF input no longer causes a provider 400 and terminates the agent loop.

- Preserves the source basename from read_file through neutral history.
- Adds the request-time media.pdf.enabled model behavior and disables it in the built-in Codex alias.
- Emits a truthful model-visible notice instead of input_file when native PDF input is disabled.
- Emits valid input_file data with a deterministic filename for PDF-capable Responses endpoints.
- Prevents parallel tool-result media duplication and rejects aggregate inline PDF payloads above the documented 50 MB limit before network submission.

## Dive Deeper

The active continuation path previously mapped every PDF MediaBlock to an OpenAI Responses input_file, even for the private Codex endpoint, which does not accept that content type. The same path discarded the source filename before serialization.

This change keeps provider compatibility at the serialization boundary:

- processMediaFile records the source basename as inlineData.displayName.
- Legacy-part conversion maps that value to MediaBlock.filename.
- media.pdf.enabled follows invocation ephemeral → model behavior → settings precedence and defaults to enabled for existing public Responses behavior.
- The Codex alias explicitly sets the capability to false.
- Disabled PDFs become input_text notices that state the PDF was not read, identify the file, and suggest extracting text or rendering pages.
- Enabled PDFs always carry a non-empty filename, falling back to document.pdf for legacy data.
- The active builder emits media once per flattened parallel tool turn rather than once per tool response.
- Aggregate validation counts only emitted inline base64 file parts and reports exact actual/allowed byte counts above 50,000,000 bytes.

The bounded delivery plan and review triage are recorded in project-plans/issue-2608-codex-pdf-input.md.

## Reviewer Test Plan

1. Run the focused provider/core/tools tests:

       cd packages/providers
       npx vitest run src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts src/openai-responses/__tests__/OpenAIResponsesProvider.pdf.test.ts src/runtime/ephemeralSettings.mediaPdf.test.ts src/utils/mediaUtils.test.ts src/composition/providerAliases.codex.test.ts

       cd ../core
       npx vitest run src/utils/generateContentResponseUtilities.test.ts

       cd ../tools
       npx vitest run src/utils/fileUtils.test.ts

2. With the Codex alias selected, ask the model to inspect multiple local PDF files in parallel. Confirm the continuation reaches the model, contains one explicit not-read notice per PDF, and does not send input_file.
3. With a PDF-capable public OpenAI Responses configuration, inspect a local PDF and confirm the outgoing request contains a PDF data URI and the source filename.
4. Run the complete local gates:

       npm run test
       npm run lint
       npm run typecheck
       npm run format
       npm run build

5. Smoke tests completed with both the ollamakimi and stepfun-37 profiles.

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  | ❓  |
| npx      | ✅  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs

Fixes #2608

### CODE REVIEW

File: packages/providers/src/openai-responses/__tests__/OpenAIResponsesProvider.pdf.test.ts [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+): Promise<Record<string, unknown>> {
+  let capturedBody: string | undefined;
+  mockFetch.mockImplementation(
+    async (
+      _input: RequestInfo | URL,
+      init?: RequestInit,
+    ): Promise<Response> => {
+      if (init?.body instanceof Blob) {
+        capturedBody = await init.body.text();
+      }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-responses/__tests__/OpenAIResponsesProvider.pdf.test.ts:86]: Evaluated against the actual request transport. openAIResponsesHttpStream.ts constructs the request body as a Blob before calling fetch, so this test intentionally asserts that concrete production contract. Handling string or ReadableStream would mask a transport regression rather than improve cross-platform behavior. CI passed this test on macOS and Linux. No code change.

File: packages/providers/src/runtime/ephemeralSettings.mediaPdf.test.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  });
+
+  describe('Codex alias pins media.pdf.enabled=false', () => {
+    it('sets media.pdf.enabled=false in ephemeralSettings', () => {
+      const aliases = loadProviderAliasEntries();
+      const codexAlias = aliases.find((a) => a.alias === 'codex');
+
+      expect(codexAlias?.config.ephemeralSettings['media.pdf.enabled']).toBe(
+        false,
+      );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/runtime/ephemeralSettings.mediaPdf.test.ts:54]: Evaluated against the parsed alias contract. ProviderAliasEntry requires config, and this test deliberately fails fast if the required config object is absent. ephemeralSettings is optional in the general config type, but Codex is required by this test to define it; adding optional chaining would make the assertion less direct. No code change.

File: packages/providers/src/utils/mediaUtils.test.ts [hunk 3]
```diff
@@ -338,3 +340,99 @@ ${webpBase64.slice(8)}
...
+
+  it('returns 0 for an empty payload', () => {
+    expect(inlineBase64ByteLength('data:application/pdf;base64,')).toBe(0);
+  });
+
+  it('computes correct length for unpadded base64', () => {
+    const padded = Buffer.from('abc').toString('base64');
+    const data = `data:application/pdf;base64,${padded.endsWith('=') ? padded.slice(0, -1) : padded}`;
+    expect(inlineBase64ByteLength(data)).toBe(3);
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/utils/mediaUtils.test.ts:438]: Fixed in commit 173490596. The test now encodes a two-byte payload that normally has one padding character, explicitly removes that character, and asserts the decoded length remains two bytes.

File: packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+
+    function pdfOfSizeBytes(bytes: number, filename: string): MediaBlock {
+      return {
+        type: 'media',
+        mimeType: 'application/pdf',
+        data: Buffer.from('A'.repeat(bytes)).toString('base64'),
+        encoding: 'base64',
+        filename,
+      };
+    }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts:369]: Evaluated and retained. The production guard scans the actual serialized data URI, so a realistic exact 50 MB boundary requires materializing the wire payload. The helper already avoids an additional decoded Buffer during production validation, and the focused test plus macOS/Linux CI passed without memory failures. Replacing this with a mock would stop testing the boundary behavior. No code change.

**Comment by acoliver** [comment]
PR OCR overflow disposition:

- OpenAIResponsesInputBuilder.ts / hasMatchingToolResponse: rejected as factually incorrect. The helper is called by appendAssistantInput when deciding whether to emit each function_call (current source around lines 201-204), so it is not dead code.
- OpenAIResponsesInputBuilder.pdf.test.ts / aggregate error assertion: fixed in commit 173490596. The assertion now derives both values from overBytes and MAX_BYTES and checks the production message order. The neighboring test separately verifies each semantic component.
- Changed-file coverage warning: no code change. This is a review-tool coverage metric rather than product test coverage. The PR adds behavioral evidence at file utility, neutral-history conversion, settings/alias, active builder, and provider request-body seams; the full local suite and macOS/Linux CI passed.

All four inline OCR findings were also evaluated, replied to individually, and resolved.

File: packages/providers/src/utils/mediaUtils.test.ts [hunk 5]
```diff
@@ -338,3 +340,100 @@ ${webpBase64.slice(8)}
...
+    expect(inlineBase64ByteLength('data:application/pdf;base64,')).toBe(0);
+  });
+
+  it('computes correct length for unpadded base64', () => {
+    const unpadded = Buffer.from('ab').toString('base64').slice(0, -1);
+    expect(
+      inlineBase64ByteLength(`data:application/pdf;base64,${unpadded}`),
+    ).toBe(2);
+  });
+});
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/utils/mediaUtils.test.ts:443]: Fixed in commit 799b76de0. Added a RED/GREEN behavioral test using an uppercase BASE64 encoding token, and made the inline data-URI matcher case-insensitive so valid mixed-case encoding tokens cannot bypass the aggregate PDF limit.

File: packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+        buildContext({ mediaPdfEnabled: false }),
+      );
+      expect(
+        allUserParts(input).filter((p) => p.type === 'input_image'),
+      ).toHaveLength(1);
+    });
+  });
+
+  describe('aggregate 50 MB native PDF preflight', () => {
+    const MAX_BYTES = 50_000_000;
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts]: Fixed in commit 799b76de0. The builder boundary tests now import PDF_AGGREGATE_MAX_BYTES from mediaUtils rather than duplicating the production limit, eliminating drift while preserving exact-boundary behavioral coverage.

File: packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+    });
+
+    it('does not emit orphan media when no valid function_call_output exists', () => {
+      const input = buildOpenAIResponsesInput(
+        [
+          {
+            speaker: 'tool',
+            blocks: [
+              toolResponse('call_orphan'),
+              pdfMediaBlock({ filename: 'orphan.pdf' }),
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts:320]: This is already covered by the dedicated aggregate-limit test near the end of the same describe block: does not count oversized PDFs in orphan tool content uses pdfOfSizeBytes(PDF_AGGREGATE_MAX_BYTES + 1024, orphan.pdf) and asserts the build does not throw. The earlier test at this thread verifies the separate serialization behavior that orphan media is not emitted. No duplicate change was made.

File: packages/providers/src/utils/mediaUtils.ts [hunk 8]
```diff
@@ -121,3 +121,43 @@ export function normalizeMediaToDataUri(media: MediaBlock): string {
...
+  const normalized = dataMatch[1].replaceAll(/\s/g, '');
+  if (normalized === '') return 0;
+  let padding = 0;
+  if (normalized.endsWith('==')) {
+    padding = 2;
+  } else if (normalized.endsWith('=')) {
+    padding = 1;
+  }
+  return Math.floor((normalized.length * 3) / 4) - padding;
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/utils/mediaUtils.ts:163]: No change. inlineBase64ByteLength is a byte-accounting helper, not a data-URI validator; malformed payload validation is outside this issue and the proposed return-0 behavior would further undercount malformed data rather than enforce the 50 MB limit. The helper intentionally returns zero only for content that is not inline base64, while normal request/provider validation remains responsible for payload validity.

File: packages/providers/src/runtime/ephemeralSettings.mediaPdf.test.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
+/**
+ * @license
+ * Copyright 2025 Vybestack LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { describe, it, expect, vi } from 'vitest';
+
+vi.unmock('./providerAliases.js');
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/runtime/ephemeralSettings.mediaPdf.test.ts]: Fixed in commit c1148faec. The unmock now targets the actual imported composition module at ../composition/providerAliases.js. Focused tests, ESLint, typecheck, formatting, and diff checks pass.

File: packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+              blocks: [
+                toolResponse('call_orphan'),
+                pdfOfSizeBytes(PDF_AGGREGATE_MAX_BYTES + 1024, 'orphan.pdf'),
+              ],
+            },
+          ],
+          buildContext(),
+        ),
+      ).not.toThrow();
+    });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts:513]: No change. This shares the same incorrect premise: orphan tool_response blocks fail hasMatchingToolCall, so no user input_file is emitted and enforcePdfAggregateLimit has nothing to count. The oversized orphan regression test passes directly on exact head c1148faec.

File: packages/providers/src/utils/mediaUtils.test.ts [hunk 11]
```diff
@@ -338,3 +340,104 @@ ${webpBase64.slice(8)}
...
+    const media: MediaBlock = {
+      type: 'media',
+      mimeType: 'application/pdf',
+      data: 'AAA',
+      encoding: 'base64',
+      filename: 'invoice.pdf',
+    };
+    const notice = buildPdfDisabledNotice(media);
+    expect(notice).toContain('invoice.pdf');
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/utils/mediaUtils.test.ts:355]: No change. The filename is model-visible data from a local tool result, not executable syntax; quotes or newlines remain inside the ordinary input_text string and cannot alter the structured Responses request. Rewriting filenames would also make the notice less truthful. Control-character normalization would be a separate display-policy feature outside issue 2608, not a security fix in PDF serialization.

File: packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+            blocks: [
+              toolResponse('call_orphan'),
+              pdfMediaBlock({ filename: 'orphan.pdf' }),
+            ],
+          },
+        ],
+        buildContext(),
+      );
+      expect(filesFrom(input)).toHaveLength(0);
+    });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/providers/src/openai-responses/__tests__/OpenAIResponsesInputBuilder.pdf.test.ts:327]: No change. The premise is incorrect: emittedOutput becomes true only after a tool_response passes hasMatchingToolCall at OpenAIResponsesInputBuilder.ts lines 240-259. In the orphan-only turn, the response is dropped, emittedOutput remains false, and the media branch at lines 266-269 is not entered. The named test passes directly on exact head c1148faec.

**Comment by acoliver** [comment]
Exact-head OCR summary disposition for c1148faec:\n\n- Optional-chain suggestion: no change. ProviderAliasEntry requires config, and the test intentionally fails fast if Codex lacks its required config or ephemeral setting. This matches the prior resolved thread on the same expression.\n- fileUtils.ts no-issues item: acknowledged; no action.\n- Changed-file coverage warning: no code change. This is an OCR sampling metric, not product test coverage; issue behavior has focused behavioral tests plus full local and hosted CI coverage.
